Move prologue_supported_inputs computations to def_kernal

laithsakka opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150869

This avoid replaying load_input on a cache hit on the generate_code_cache.
the idea is that if a template have prologue_loads_all_inputs = True, it means that 
all all inputs are loaded and hence no need to replay 

Effect on the current benchmark on a local run on dev server.
18549985383 -> 15072230073
25697270062 -> 20738613297

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

laithsakka committed:
move prologue_supported_inputs computations to def_kernal

[ghstack-poisoned]

laithsakka committed:
Update on "move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
23319257037 -> 19882931330
31341716252 -> 27035999565

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
23319257037 -> 19882931330
31341716252 -> 27035999565

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
20890757277 -> 17471274163
28316151238 -> 24025543849

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
20890757277 -> 17471274163
28316151238 -> 24025543849

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
20890757277 -> 17471274163
28316151238 -> 24025543849

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
20890757277 -> 17471274163
28316151238 -> 24025543849

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
effect on the current benchmark:
20890757277 -> 17471274163
28316151238 -> 24025543849

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

eellison commented on torch/_inductor/kernel/mm.py:
```diff
@@ -223,6 +223,7 @@
 """
     ),
     cache_codegen=True,
+    all_inputs_loaded=True,
```
nit: maybe `prologue_loads_all_inputs`

eellison commented on torch/_inductor/select_algorithm.py:
```diff
@@ -752,7 +759,9 @@ def load_input(
 
         input_node = self.named_input_nodes[input_name]
-        self.prologue_supported_inputs.add(input_node.get_name())
+        if not self.all_inputs_loaded:
+            self.prologue_supported_inputs.add(input_node.get_name())
+
```
This is an OrderedSet, so we dont need this check. If we're going to do this we should insert that its already in the set

drisspg commented on torch/_inductor/select_algorithm.py:
```diff
@@ -385,6 +386,8 @@ def __init__(
         # Update each time an input is marked frozen.
         self.frozen_layouts_cnt = 0
 
+        self.all_inputs_loaded = all_inputs_loaded
```
maybe a quick description of what this does?

laithsakka committed:
Update on "Move prologue_supported_inputs computations to def_kernal"


This avoid replaying load_input on a cache hit on the generate_code_cache.
the idea is that if a template have prologue_loads_all_inputs = True, it means that 
all all inputs are loaded and hence no need to replay 

Effect on the current benchmark on a local run on dev server.
18549985383 -> 15072230073
25697270062 -> 20738613297

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

laithsakka commented:
@pytorchbot merge
